Tests/ConnectionPoolModuleTests/ConnectionPoolTests.swift

```diff
@@ -368,6 +368,65 @@ final class ConnectionPoolTests: XCTestCase {
         }
     }
 
-}
+    func testCancelConnectionRequestWorks() async throws {
+        let clock = MockClock()
+        let factory = MockConnectionFactory<MockClock>()
+        let keepAliveDuration = Duration.seconds(30)
+        let keepAlive = MockPingPongBehavior(keepAliveFrequency: keepAliveDuration, connectionType: MockConnection.self)
+
+        var mutableConfig = ConnectionPoolConfiguration()
+        mutableConfig.minimumConnectionCount = 0
+        mutableConfig.maximumConnectionSoftLimit = 4
+        mutableConfig.maximumConnectionHardLimit = 4
+        mutableConfig.idleTimeout = .seconds(10)
+        let config = mutableConfig
 
+        let pool = ConnectionPool(
+            configuration: config,
+            idGenerator: ConnectionIDGenerator(),
+            requestType: ConnectionRequest<MockConnection>.self,
+            keepAliveBehavior: keepAlive,
+            observabilityDelegate: NoOpConnectionPoolMetrics(connectionIDType: MockConnection.ID.self),
+            clock: clock
+        ) {
+            try await factory.makeConnection(id: $0, for: $1)
+        }
+
+        try await withThrowingTaskGroup(of: Void.self) { taskGroup in
+            taskGroup.addTask {
+                await pool.run()
+            }
+
+            let leaseTask = Task {
+                do {
+                    _ = try await pool.leaseConnection()
+                    XCTFail("Did not expect call to succeed")
+                } catch {
+                    XCTAssertEqual((error as? ConnectionPoolError), .requestCancelled)
```

Comment:
Haven't the XCTAssert...Async utilities (either my version or NIO's (mine's better, of course)) shown up by now, instead of these awkward inexpressive-of-the-actual-intent constructs? (I mean, `swift-testing`'s gonna utterly blast this sorta thing outta the water anyway, of course...) Here, have the full set: https://gist.github.com/gwynne/2fbbc57eaad942368ea87dd0d2e6628e

Reply:
Not sure this comment still applies after your revisions, but I'll leave it here just in case.

Reply:
The connection pool shall not have any weird dependencies. So we need to dance here.